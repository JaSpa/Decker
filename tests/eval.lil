
# parse error
show[eval["99`5"]]

# inappropriate keyword
show[eval["if 1 do 2 else 3 end"].error]

# empty expression
show[eval[""]]

# simple expression
show[eval["1+2"]]

# chaining extraction
show[eval["2+3"].value]

# reference passed-in bindings
foo.a:100
foo.b:200
show[eval["a+b" foo]]

# capture new bindings
show[eval["foo:69"]]

# external bindings don't leak in
show[eval["foo"]]

# internal bindings don't leak out
c:444
show[eval["c:42 11+22"]]
show[c]

# passed-in binding dict doesn't mutate
show[eval["c:999" foo]]
show[foo]

# parser edge case: ending on a keyword
show[eval["on foo do 99 end"]]

# functions built in an eval[] retain the correct closure
inside:eval["on one x do 100*x end  on two x do one[3*x] end"].value
show[inside[5]]

outside.const:300
inside:eval["on one x do const*x end  on two x do one[3+x] end" outside].value
show[inside[4]]
