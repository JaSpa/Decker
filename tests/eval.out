{"value":0,"vars":{},"error":"Invalid character '`'.","errorpos":(1,3)}
"'do' is a keyword, and cannot be used for a variable name."
{"value":0,"vars":{"__proto__":0}}
{"value":0,"vars":{}}
{"value":3,"vars":{}}
5
{"value":300,"vars":{"a":100,"b":200}}
{"value":69,"vars":{"foo":69}}
{"value":0,"vars":{}}
{"value":33,"vars":{"c":42}}
444
{"value":999,"vars":{"a":100,"b":200,"c":999}}
{"a":100,"b":200}
{"value":on foo do ... end,"vars":{"foo":on foo do ... end}}
1500
2100
{"value":777,"vars":{"localish":10008}} 0 777
